[TAN-5277] Data migration: legacy files → file + attachment

adessy opened the pull request:
# Changelog
## Technical
- [TAN-5277] Implements the migration logic to convert legacy files into the new `Files::File` model.
- Similar to what was done for the `WebApi::V1::FilesController` (in commit 9bfa59b),  refactors `WebApi::V1::IdeasController` to support `FileAttachment` alongside the legacy `FileUpload`.

adessy committed:
[TAN-5277] Add `migrated_file_id` to mark legacy files that have been migrated

The default_scope of the legacy file model is adapted to exclude the files marked as migrated.

adessy committed:
[TAN-5277] Update `update_file_upload_fields` to support `FileAttachment` alongside legacy `FileUpload`

Similar to what was done for the `WebApi::V1::FilesController` (in commit 9bfa59bcb4657d70aa958e220425cd3254c8f8ca), this PR refactors `update_file_upload_fields` to support `FileAttachment` alongside the legacy `FileUpload`, easing the migration from legacy files (including `FileUpload`) to `FileAttachment` plus `Files::File`.

adessy commented on back/app/controllers/web_api/v1/ideas_controller.rb:
```diff
@@ -376,28 +376,48 @@ def extract_params_for_file_upload_fields(custom_form, params)
     params_for_file_upload_fields.each do |key, params_for_files_field|
-      if params_for_files_field['id']
-        idea_file = FileUpload.find(params_for_files_field['id'])
-        if idea_file
```
Remove the condition that is never `false`, since `FileUpload.find` raises an exception if the record is not found.

adessy committed:
[TAN-5277] Adapt various input visitor to use both file_ideas and file_attachments

adessy committed:
[TAN-5277] Rework includes for idea files

Some of them were not necessary. In other cases, we added `attached_files` and `file_attachments` to the include list.

adessy committed:
[TAN-5277] Add `LegacyFileMigrationService` to handle migration of legacy files to the new unified file system

adessy committed:
[TAN-5277] Add logging and migration statistics to `LegacyFileMigrationService`

adessy committed:
[TAN-5277] Enhances a legacy file migration service

- Add methods to revert the migration
- Add support for multi-tenancy
- Various improvements in code quality

adessy committed:
[TAN-5277] Add `migration_skipped_reason` field to legacy files

adessy committed:
[TAN-5277] Add tests for `Files::LegacyFileMigrationService`

kogre commented on back/engines/commercial/analysis/app/controllers/analysis/web_api/v1/inputs_controller.rb:
Could there be a front-end <> back-end  N+1 issue on file_attachments? I guess we were including idea_files for a reason?

kogre reviewed (approved):
Looks good! Went through the code (not in detail through the specs) and wasn't able to spot any obvious weirdness or mistakes.

Must say that I am not 100% following how the transition is affecting the front-end, that's maybe my main worry, but I assume this is fully covered and tested?

Maybe we can do a test run on a sandbox cluster and do some manual tests?

adessy committed:
[TAN-5277] Fix rubocop offenses

adessy committed:
Merge branch 'master' into TAN-5277/data-migration-legacy-files

# Conflicts:
#	back/db/structure.sql

adessy commented on back/app/controllers/web_api/v1/ideas_controller.rb:
```diff
@@ -66,7 +66,7 @@ def index_idea_markers
     ).find_records
     ideas = paginate SortByParamsService.new.sort_ideas(ideas, params, current_user)
-    ideas = ideas.includes(:author, :topics, :project, :idea_status, :idea_files)
+    ideas = ideas.includes(:author, :topics, :project, :idea_status)
```
@kogre, I think we could consider adding this one back. I don't immediately see why it is included for the idea markers but not when indexing ideas, except maybe because the user can browse ideas on the map without leaving the page.

I think the other changes to the includes look fine.

adessy replied:
Actually, I think I was right. I’m not sure why they are included. They don’t seem to be used by the serializer, nor are they included in the JSON API response.

adessy committed:
[TAN-5277] Remove test for file inclusion in AI Analysis inputs

Files have been removed from the included resources because they are not used on the front-end side.

adessy committed:
[TAN-5277] Fix tests broken by an incorrect merge

adessy committed:
[TAN-5277] Exclude deleted tenants from legacy file migration

Since migrations are not run on deleted tenants, their schema gets out of sync and causes errors.
